Add instructions for NAOT cross-compile (#37428)
dotnet/docs: docs/core/deploying/native-aot/cross-compile.md

---
title: Cross-compilation
description: Instructions and limitations for cross-compilation
author: agocke
ms.author: angocke
ms.date: 10/10/2023
---

# Cross-compilation

Due to restrictions of the platforms themselves, **Native AOT does not support cross-OS

> compilation**, meaning you cannot directly compile for Windows using Linux or vice versa. Cross-OS

Reviewer: ```suggestion
compilation**. That means you can't directly compile for Windows using Linux or vice versa. Cross-OS
```

compilation requires some form of emulation, like a virtual machine or Windows WSL.

However, Native AOT does have limited support for _cross-architecture_ compilation. As long as the
necessary native toolchain is installed, you can cross-compile between the `x64` and the `arm64`
architectures on Windows, Mac, or Linux.

## Windows

> Windows will require the toolchain for both the host and target architecture. This means you will need

Reviewer: ```suggestion
Windows requires the toolchain for both the host and target architecture. This means you need
```

the MSVC C/C++ Visual Studio workloads for both x64/x86 and for arm64.

## Mac

MacOS provides the x64 and amd64 toolchains in the default XCode install.

## Linux

> Every Linux distribution has a different system for installing native toolchain dependencies. The documentation for your Linux distribution will need to be consulted to determine the necessary steps.

Reviewer: ```suggestion
Every Linux distribution has a different system for installing native toolchain dependencies. Consult the documentation for your Linux distribution to determine the necessary steps.
```

The necessary dependencies are:

> - A cross-linker, or a linker which can emit for the target. `clang` is one such linker

Reviewer: ```suggestion
- A cross-linker, or a linker that can emit for the target. `clang` is one such linker.
```

> - A target-compatible `objcopy` or `strip`, if `StripSymbols` is enabled for your project

Reviewer: ```suggestion
- A target-compatible `objcopy` or `strip`, if `StripSymbols` is enabled for your project.
```

> - Object files for the C runtime of the target architecture

Reviewer: ```suggestion
- Object files for the C runtime of the target architecture.
```

> - Object files for zlib for the target architecture

Reviewer: ```suggestion
- Object files for zlib for the target architecture.
```
